---
title: "h) Downloading DNA sequences"
author: "Dr. Avril Coghlan, adapted by Nathan Brouwer"
date: "`r Sys.Date()`"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{Vignette Title}
  %\VignetteEncoding{UTF-8}
  %\VignetteEngine{knitr::rmarkdown}
editor_options: 
  chunk_output_type: console
---

```{r setup, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>",
  warnings = FALSE
)
```

This is a modification of ["DNA Sequence Statistics"](https://a-little-book-of-r-for-bioinformatics.readthedocs.io/en/latest/src/chapter1.html) from Avril Coghlan's [*A little book of R for bioinformatics.*](https://a-little-book-of-r-for-bioinformatics.readthedocs.io/en/latest/index.html).  Almost all of text and code was originally written by Dr. Coghlan and distributed under the [Creative Commons 3.0](https://creativecommons.org/licenses/by/3.0/us/) license.  


### Functions

library()
help()
length
table
seqinr::GC
seqinr::count
seqinr::write.fasta

### Software/websites

www.ncbi.nlm.nih.gov
Text editor (eg Notepad++, TextWrangler)

### R vocabulary

list
library
package
CRAN
wrapper


### File types

FASTA

### Bioinformatics vocabulary

accession
NCBI
NCBI Sequence Database
EMBL Sequence Database
FASTA file

### Organisms and Sequence accessions

Dengue virus: DEN-1, DEN-2, DEN-3, and DEN-4. The NCBI accessions for the DNA sequences of the DEN-1, DEN-2, DEN-3, and DEN-4 Dengue viruses are NC_001477, NC_001474, NC_001475 and NC_002640, respectively.


## DNA Sequence Statistics: Part 1

### Using R for Bioinformatics

This booklet tells you how to use the R software to carry out some simple analyses that are common in bioinformatics. In particular, the focus is on computational analysis of biological sequence data such as genome sequences and protein sequences.

This booklet assumes that the reader has some basic knowledge of biology, but not necessarily of bioinformatics. The focus of the booklet is to explain simple bioinformatics analysis, and to explain how to carry out these analyses using *R*.

To use *R*, you first need to start the *R* program on your computer. You should have already installed *R* on your computer (if not, for instructions on how to install R, see earlier tutorials in this package).

### R packages for bioinformatics: Bioconductor and SeqinR

Many authors have written *R* packages for performing a wide variety of analyses. These do not come with the standard *R* installation, but must be installed and loaded as “add-ons”.

Bioinformaticians have written several specialised packages for R. In this practical, you will learn to use the [*SeqinR*](https://cran.r-project.org/web/packages/seqinr/index.html) package to retrieve sequences from a DNA sequence database, and to carry out simple analyses of DNA sequences.

Some well-known bioinformatics packages for *R* are the Bioconductor set of R packages (www.bioconductor.org), which contains several packages with many *R* functions for analysing biological data sets such as microarray data; and the [*SeqinR*](https://cran.r-project.org/web/packages/seqinr/index.html) package, which contains R functions for obtaining sequences from DNA and protein sequence databases, and for analysing DNA and protein sequences.  

To use functions from the *SeqinR* package, we first need to install the *SeqinR* package (for instructions on how to install an R package, see [How to install an R package](https://a-little-book-of-r-for-bioinformatics.readthedocs.io/en/latest/src/installr.html)).

We will also use functions from the *rentrez* and *ape* packages.


The easiest way to install a package is to use the install.packages() command:
```{r, echo = F, eval = F}
install.packages("seqinr")
install.packages("rentrez")
install.packages("ape")
```


Once you have installed a R package, you can load it using the library() command

```{r, echo = F}
library("seqinr")
library("rentrez")
library("ape")
```

You may see some angry-looking red text pop up.  No worries.

Remember that you can ask for more information about a particular R command by using the `help() eval = F` function. For example, to ask for more information about the `library() eval = F function`, you can type:

```{r, eval = F}
help("library")
```



### FASTA format

The FASTA format is a simple and widely used format for storing biological (DNA or protein) sequences. It was first used by the FASTA program for sequence alignment. It begins with a single-line description starting with a “>” character, followed by lines of sequences. Here is an example of a FASTA file:
```{r, eval = T, echo = F}
cat("A06852 183 residues MPRLFSYLLGVWLLLSQLPREIPGQSTNDFIKACGRELVRLWVEICGSVSWGRTALSLEEPQLETGPPAETMPSSITKDAEILKMMLEFVPNLPQELKATLSERQPSLRELQQSASKDSNLNFEEFKKIILNRQNEAEDKSLLELKNLGLDKHSRKKRLFRMTLSEKCCQVGCIRKDIARLC")
```


### The NCBI sequence database

The [National Centre for Biotechnology Information (NCBI)](www.ncbi.nlm.nih.gov) in the US maintains a huge database of all the DNA and protein sequence data that has been collected, the NCBI Sequence Database. This also a similar database in Europe, the [European Molecular Biology Laboratory (EMBL) Sequence Database](www.ebi.ac.uk/embl), and also a similar database in Japan, the [DNA Data Bank of Japan (DDBJ)](www.ddbj.nig.ac.jp). These three databases exchange data every night, so at any one point in time, they contain almost identical data.

Each sequence in the NCBI Sequence Database is stored in a separate record, and is assigned a unique identifier that can be used to refer to that sequence record. The identifier is known as an **accession**, and consists of a mixture of numbers and letters. For example, Dengue virus causes Dengue fever, which is classified as a neglected tropical disease by the WHO, is classified by any one of four types of Dengue virus: DEN-1, DEN-2, DEN-3, and DEN-4. The NCBI accessions for the DNA sequences of the DEN-1, DEN-2, DEN-3, and DEN-4 Dengue viruses are NC_001477, NC_001474, NC_001475 and NC_002640, respectively.

Note that because the NCBI Sequence Database, the EMBL Sequence Database, and DDBJ exchange data every night, the DEN-1 (and DEN-2, DEN-3, DEN-4) Dengue virus sequence will be present in all three databases, but it will have different accessions in each database, as they each use their own numbering systems for referring to their own sequence records.

### Retrieving genome sequence data via the NCBI website

You can easily retrieve DNA or protein sequence data from the NCBI Sequence Database via its website www.ncbi.nlm.nih.gov.

The Dengue DEN-1 DNA sequence is a viral DNA sequence, and as mentioned above, its NCBI accession is NC_001477. To retrieve the DNA sequence for the Dengue DEN-1 virus from NCBI, go to the NCBI website, type “NC_001477” in the Search box at the top of the webpage, and press the “Search” button beside the Search box:

( image0 unvailable )

On the results page you will see the number of hits to “NC_001477” in each of the NCBI databases on the NCBI website. There are many databases on the NCBI website, for example, the “PubMed” data contains abstracts from scientific papers, the “Nucleotide” database contains DNA and RNA sequence data, the “Protein” data contains protein sequence data, and so on. The picture below shows what the results page should look like for your NC_001477 search. As you are looking for the DNA sequence of the Dengue DEN-1 virus genome, you expect to see a hit in the NCBI Nucleotide database, and indeed there is hit in the Nucleotide database (indicated by the “1” beside the icon for the Nucleotide database):

( image1 unvailable)

To look at the one sequence found in the Nucleotide database, you need to click on the icon for the NCBI Nucleotide database on the results page for the search:

( image2 unvailable)

When you click on the icon for the NCBI Nucleotide database, it will bring you to the record for NC_001477 in the NCBI Nucleotide database. This will contain the name and NCBI accession of the sequence, as well as other details such as any papers describing the sequence:

( image3 unvailable)

To retrieve the DNA sequence for the DEN-1 Dengue virus genome sequence as a FASTA format sequence file, click on “Send” at the top right of the NC_001477 sequence record webpage, and then choose “File” in the pop-up menu that appears, and then choose FASTA from the “Format” menu that appears, and click on “Create file”.

A box will pop up asking you what to name the file, and where to save it. You should give it a sensible name (eg. “den1.fasta”) and save it in a place where you will remember (eg. in the “My Documents” folder is a good idea):

( image4 )

You can now open the FASTA file containing the DEN-1 Dengue virus genome sequence using WordPad (***or another text editor such as Notepad++) on your computer. To open WordPad, click on “Start” on the bottom left of your screen, click on “All Programs” in the menu that appears, and then select “Accessories” from the menu that appears next, and then select “WordPad” from the menu that appears next. WordPad should start up. In Wordpad, choose “Open” from the “File” menu. The WordPad “Open” dialog will appear. Set “Files of type” to “All Documents” at the bottom of the WordPad “Open” dialog. You should see a list of files, now select the file that contains the DEN-1 Dengue virus sequence (eg. “den1.fasta”). The contents of the FASTA format file containing the Dengue DEN-1 sequence should now be displayed in WordPad:

( image5 )

### Retrieving genome sequence data using rentrez

Instead of going to the NCBI website to retrieve sequence data from the NCBI database, you can retrieve sequence data from NCBI directly from *R*, by using the *rentrez* package.

For example, you learnt above how to retrieve the DEN-1 Dengue virus genome sequence, which has NCBI accession NC_001477, from the NCBI website. To retrieve a sequence with a particular NCBI accession, you can use an R function entrez_fetch() from the rentrez package.

```{r}
dengueseq_fasta <- entrez_fetch(db = "nucleotide", 
                          id = "NC_001477", 
                          rettype = "fasta")
```

Ok, so what exactly have we done?  We have an object "dengueseq_fasta" which has the sequence linke to the accession number "NC_001477."  So where is the sequence?

If we want to do it we can type just type "dengueseq_fasta" and press enter

```{r, echo = F}
strtrim(dengueseq_fasta,500)
```

Note that I have only shown the first 500 of the over 10,000 bases in the sequence.

If you scroll back up to the type of all the bases you'll see the identifier of the sequence

```{r, echo = F}
strtrim(dengueseq_fasta, 44)
```

Note that at the end of teh name is a "\n" which indicates to the computer that this is a newline; this is read by texteditor, but is ignored by R in this contxt.
```{r, echo = F}
strtrim(dengueseq_fasta, 45)
```

After the "\n" will beging the sequence, which will continue on for a LOOOOOONG way.
```{r, echo = F}
strtrim(dengueseq_fasta, 52)
```

dengueseq is an are object.  We can find out more information about what it is using the class() command

```{r}
class(dengueseq_fasta)
```

Many things in R are vectors so we can ask R is.vector()

```{r}
is.vector(dengueseq_fasta)
```

Yup, that's true.

Ok, let's see what else.  A handy though often verbose command is is(), which tells us what an object, well, what it is:

```{r}
is(dengueseq_fasta)
```

There is a lot here but if you scan for some key words you will see "character" and "vector" at the top.  The otherstuff you can ignore.  The first two things, though, tell us the dengueseq_fasta is a **vector** of the class **character**: a **character vector**.

Another handy function is str(), which gives us a peak at the context and structure of an R object.
```{r}
str(dengueseq_fasta)
```

We know it contains character data - how man?  nchar() for "number of characters" answers that:
```{r}
nchar(dengueseq_fasta)
```

Over 10,000 character, most of the A, T, C, and G.

One last thing: length().  I'm not actually going to explain this right now, but thing about why the answer is what it is, and why its so different from what nchar() told us:
```{r}
length(dengueseq_fasta)
```

Hint: how many pieces of unque information are in the dengueseq object?  In what sense is there only 1 piece of information?


## Convert FASTA to an R variable

We can't actuall do much with the contents of the dengueseq_fasta we downloaded with the rentrez package.  What we need is to convert it into something R can work with.  
There are several things we need to take care of. 

1. The meta data ">NC_001477.1 Dengue virus 1, complete genome"
1. All the "\n" that show up in the file.

The second item is actually easier to take care of.   R and many programming languages have tools called "regular expressions" that allow you to manipulate text.  R has a function called gsub() which allows you to substitute or delete character data from a string.  Firt I'll remove all those n values.

```{r}
dengueseq_vector <- gsub("\n", "", dengueseq_fasta)
```

We can use str() to see if it worked

```{r}
str(dengueseq_vector)
```

Now for the metadata.  This is a bit complex, but the following code is going to talk all the that occurs before the beginning of the sequnce ("AGTTGTTAGTC") and delete it.


```{r}
header. <- ">NC_001477.1 Dengue virus 1, complete genome"
dengueseq_vector <- gsub(header.,"", dengueseq_vector)
```

```{r}
str(dengueseq_vector)
```


```{r}
dengueseq_vector <- stringr::str_split(dengueseq_vector,pattern = "",
                                       simplify = FALSE)[[1]]
```


```{r}
str(dengueseq_vector)
```


Ok, that was a lot of work, so I've packaged all of that up into a function for future use


Something cool which we will explore in the next exercise is that we can do summaries on vectors of nucleotides, like this
```{r}
table(dengueseq_vector)
```


## Saving FASTA files

We can save our data as .fasta file for safe keeping

```{r, eval = F}
write(dengueseq_fasta, file="dengueseq.fasta")
```

## Reading in FASTA files

We can read in FASTA files

```{r, echo = F}
fi. <- here::here("inst/extdata/dengueseq.fasta")
# dengueseq <- seqinr::read.fasta(file = fi.,
#                              seqtype  = "DNA",
#                              as.string = FALSE)

dengueseq_matrix <- ape::read.dna(file = here::here("inst/extdata/dengueseq.fasta"),
              format = "fasta",
              as.character = TRUE)
```


```{r, eval = F}
dengueseq_matrix <- ape::read.dna(file = system.file("dengueseq.fasta"),
              format = "fasta",
              as.character = TRUE)
```


Note that reading it back in like this results in a change from its original format

```{r}
class(dengueseq_matrix)
```

Now we have a matrix instead of a vector.  str() will give us a peak at it
```{r}
str(dengueseq_matrix)
```

Matrices are squares of data - how big is this quare?

```{r}
dim(dengueseq_matrix)
```

We have a 1 x 10737 matrix.

To print out a certain subsequence of the sequence, we just need to type the name of the vector dengueseq_reload followed by the square brackets [ ] containing the **indices** for the nucleotides we want to see. For example, the following command prints out the first 50 nucleotides of the DEN-1 Dengue virus genome sequence:

```{r}
dengueseq_matrix[1:50]
```

Note that dengueseq_matrix[1:50] refers to the elements of the vector dengueseq_matrix with indices from 1-50. These elements contain the first 50 nucleotides of the DEN-1 Dengue virus sequence.


We can run table() on this 
```{r}
table(dengueseq_matrix)
```


```{r}

```


