---
title: "e1) Downloading sequences from UniProt by hand"
author: "Nathan Brouwer"
date: "`r Sys.Date()`"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{Vignette Title}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r setup, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>",
  warnings = FALSE
)
```



## UniProt
In the previous chapter you learnt how to retrieve DNA and protein sequences from the NCBI database. The NCBI database is a key database in bioinformatics because it contains essentially all DNA sequences ever sequenced.

As mentioned in the previous chapter, a subsection of the NCBI database called RefSeq consists of high quality DNA and protein sequence data. Furthermore, the NCBI entries for the RefSeq sequences have been manually curated, which means that biologists employed by NCBI have added additional information to the NCBI entries for those sequences, such as details of scientific papers that describe the sequences.

Another extremely important manually curated database is UniProt (www.uniprot.org), which focuses on protein sequences. UniProt aims to contains manually curated information on all known protein sequences. While many of the protein sequences in UniProt are also present in RefSeq, the amount and quality of manually curated information in UniProt is much higher than that in RefSeq.

For each protein in UniProt, the UniProt curators read all the scientific papers that they can find about that protein, and add information from those papers to the protein’s UniProt entry. For example, for a human protein, the UniProt entry for the protein usually includes information about the biological function of the protein, in what human tissues it is expressed, whether it interacts with other human proteins, and much more. All this information has been manually gathered by the UniProt curators from scientific papers, and the papers in which the found the information are always listed in the UniProt entry for the protein.

Just like NCBI, UniProt also assigns an accession to each sequence in the UniProt database. Although the same protein sequence may appear in both the NCBI database and the UniProt database, it will have different NCBI and UniProt accessions. However, there is usually a link on the NCBI entry for the protein sequence to the UniProt entry, and vice versa.

Viewing the UniProt webpage for a protein sequence
If you are given the UniProt accession for a protein, to find the UniProt entry for the protein, you first need to go the UniProt website, www.uniprot.org. At the top of the UniProt website, you will see a search box, and you can type the accession of the protein that you are looking for in this search box, and then click on the “Search” button to search for it.

For example, if you want to find the sequence for the chorismate lyase protein from Mycobacterium leprae (the bacterium which causes leprosy), which has UniProt accession Q9CD83, you would type just “Q9CD83” in the search box and press “Search”:

image0

The UniProt entry for UniProt accession Q9CD83 will then appear in your web browser. The picture below shows the top part of the UniProt entry for accession Q9CD83. You can see there is a lot of information about the protein in its UniProt entry.

Beside the heading “Organism” you can see the organism is given as Mycobacterium leprae. Beside the heading “Taxonomic lineage”, you can see “Bacteria Actinobacteria Actinobacteridae Actinomycetales Corynebacterineae Mycobacteriaceae Mycobacterium”.

This tells us that Mycobacterium is a species of bacteria, which belongs to a group of related bacteria called the Mycobacteriaceae, which itself belongs to a larger group of related bacteria called the Corynebacterineae, which itself belongs to an even larger group of related bacteria called the Actinomycetales, which itself belongs to the Actinobacteridae, which itself belongs to a huge group of bacteria called the Actinobacteria.

Beside the heading “Sequence length” we see that the sequence is 210 amino acids long (210 letters long). Further down, beside the heading “Function”, it says that the function of this protein is that it “Removes the pyruvyl group from chorismate to provide 4-hydroxybenzoate (4HB)”. This tells us this protein is an enzyme (a protein that increases the rate of a specific biochemical reaction), and tells us what is the particular biochemical reaction that this enzyme is involved in.

Further down the UniProt page for this protein, you will see a lot more information, as well as many links to webpages in other biological databases, such as NCBI. The huge amount of information about proteins in UniProt means that if you want to find out about a particular protein, the UniProt page for that protein is a great place to start.

image1

Retrieving a UniProt protein sequence via the UniProt website
To retrieve a FASTA-format file containing the sequence for a particular protein, you need to look at the top right of the UniProt entry for the protein on the UniProt website.

You will see a small orange button labelled “FASTA”, which you should click on:

image2

The FASTA-format sequence for the accession will now appear in your web browser. To save it as a file, go to the “File” menu of your web browser, choose “Save page as”, and save the file. Remember to give the file a sensible name (eg. “Q9CD83.fasta” for accession Q9CD83), and in a place that you will remember (eg. in the “My Documents” folder).

For example, you can retrieve the protein sequences for the chorismate lyase protein from Mycobacterium leprae (which has UniProt accession Q9CD83) and for the chorismate lyase protein from Mycobacterium ulcerans (UniProt accession A0PQ23), and save them as FASTA-format files (eg. “Q9CD83.fasta” and “A0PQ23.fasta”, as described above.

Note that Mycobacterium leprae is the bacterium which causes leprosy, while Mycobacterium ulcerans is a related bacterium which causes Buruli ulcer, both of which are classified by the WHO as neglected tropical diseases.

Note that the M. leprae and M. ulcerans chorismate lyase proteins are an example of a pair of homologous (related) proteins in two related species of bacteria.

Once you have downloaded the protein sequences for UniProt accessions Q9CD83 and A0PQ23 and saved them as FASTA-format files (eg. “Q9CD83.fasta” and “A0PQ23.fasta”), you can read them into R using the read.fasta() function in the SeqinR R package (as described in chapter 1).

Remember that the read.fasta() function expects that you have put your FASTA-format files in the “My Documents” folder on your computer.

For example, the following commands will read the FASTA-format files Q9CD83.fasta and A0PQ23.fasta into R, and store the two protein sequences in two vectors lepraeseq and ulceransseq:


```{r}
library("seqinr")
```

```{r}
leprae <- read.fasta(file = "Q9CD83.fasta")
```


```{r}
ulcerans <- read.fasta(file = "A0PQ23.fasta")
```

```{r}
lepraeseq <- leprae[[1]]
```


```{r}
ulceransseq <- ulcerans[[1]]
```


Display the contents of the vector "lepraeseq"
```{r}
lepraeseq 
```



